Notes/Proposals/Explicit-partial-type-inference.md

```diff
@@ -0,0 +1,313 @@
+It would help to make documentation right by the code with the meaning "This segment of code primarily cares about the wrapper".
+
+```csharp
+var a = new Wrapper<_>(wrappedElement)
+```
+
+Another need where `_` placeholders can be used is specifying the arity of type argument on places, where can be ambiguities like `IEnumerable` vs `IEnumerable<_>`.
+
+Worth to mention other options which could be accomplished in the future regarding default and named type arguments.
+Having the `_` placeholder can be used as a shortcut for choosing the right generic overload and to save typing when we use named type parameters.
+
+```csharp
+class Foo<T1, T2 = int> {}
+class Foo<T1, T2 = int, T3 = string> {}
+
+new Foo<T1: _, T2: string, T3 = _>() // Assuming that T1 can be inferred and T3 is default.
+new Foo<_,_> // Choosing Foo<T1, T2> based on the arity
+```
+
+Method type inference(including object creation) is not the only place where we can use the `_` placeholder.
+Sometimes `var` keyword as a variable declaration is not sufficient. 
+We would like to be able to specify more the type information about variable but still have some implementation details hidden.
+With the `_` placeholder we would be able to specify more the shape of the variable avoiding unnecessary specification of type arguments. 
+
+```csharp
+Wrapper<_> wrapper = ... // I get an wrapper, which I'm interested in, but I don't care about the type arguments, because I don't need them in my code.
+wrapper.DoSomething();
+```
+At the end, casting could use the `_` placeholder as well.
+
+```csharp
+Foo<int, string> myvar = (Foo<_,_>)myobject; // Hint the type arguments based on target or other potential source of type information like default or named type arguments.
+```
+
+An interesting thing would be to allow the `_` placeholder in member lookup as you can see in the example below.
+On the first see, it can look wierd.
+
+```csharp
+static class C1<T1> {
+    static class C2 <T2> {
+        public (T1, T2) Foo(T1 t, T2 t) {}
+    }
+}
+
+var a = C1<_>.C2<_>.Foo(1, 1);
+```
+
+But in combination with default parameters, It might be useful in cases, where we use entity as a global provider of something, which we determine by type.
+
+```csharp
+static class Factory<T = Default> {
+    public static T Create(){...}
+}
+
+int a = Factory<_>.Create(); // Calls Factory<int>.Create();
+var b = Factory<_>.Create(); // Calls Factory<Default>.Create();
+```
+
+Although it is unlikely that it would be added into C#, we would like to investigate what should be done to enable it for a deeper understanding of type inference in Roslyn.
+
+Based on mentioned advantages that `_` placeholder can bring to type inference we feel to add it to C# worthly.
+
+## Scope
+
+Partial type inference can be solved in various ways.
+We chose a feature enabling to hint the compiler by specifying ambitious type arguments and letting the remaining ones on him.
```

Comment:
> ambitious

Did you mean ambiguous?

> letting the remaining ones on him

It will sound more natural to say something like: *letting the compiler infer the rest*.